diff --git a/src/includes/perfmon_perfevent.h b/src/includes/perfmon_perfevent.h
index 98ad2ef6..3d2567ac 100644
--- a/src/includes/perfmon_perfevent.h
+++ b/src/includes/perfmon_perfevent.h
@@ -292,6 +292,10 @@ int parse_event_config(char* base, char* option, int* num_formats, struct perf_e
             bdestroy(src);
             fclose(fp);
         }
+        else
+        {
+            err = -ENOENT;
+        }
         bdestroy(path);
     }
     else
@@ -813,14 +817,14 @@ int perf_uncore_setup(struct perf_event_attr *attr, RegisterType type, PerfmonEv
 /*        free(formats);*/
 /*    }*/
 #ifdef _ARCH_PPC
-    if (ret == 0 && reg == INVALID)
+    if (ret == -ENOENT)
     {
         num_formats = 0;
         formats = NULL;
         ret = parse_event_config(bdata(perf_folder), perfEventOptionNames[EVENT_OPTION_UNCORE_CONFIG], &num_formats, &formats);
         if (ret == 0)
         {
-            for (int i = 0; i < num_formats && umask != 0x0; i++)
+            for (int i = 0; i < num_formats && eventConfig != 0x0; i++)
             {
                 switch(formats[i].reg)
                 {
